File: javascript/beer-song/beer-song.js
'use strict';

function format(message, params) {
  return message
    .replace(/\{0\}/g, params[0])
    .replace(/\{1\}/g, params[1]);
}

function capitalize(str) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}

function pluralize(n) {
  var suffix = (n === 1) ? '' : 's';
  return 'bottle' + suffix;
}

function bottles(n) {
  switch(n) {
    case 0: return 'no more';
    case -1: return '99';
    default: return n.toString();
  }
}

function bottlesOfBeer(n) {
  return bottles(n) + ' ' + pluralize(n);
}

function start(n) {
  var params = [ bottlesOfBeer(n) ];
  var result = format('{0} of beer on the wall, {0} of beer.', params);
  return capitalize(result);
}

function takeOneDown(n) {
  switch(n) {
    case 1: return 'Take it down and pass it around';
    case 0: return 'Go to the store and buy some more';
    default: return 'Take one down and pass it around';
  }
}

function end(n) {
  var params = [ takeOneDown(n), bottlesOfBeer(n - 1) ];
  return format('{0}, {1} of beer on the wall.', params);
}

module.exports = {

  verse: function (n) {
    return []
      .concat(start(n))
      .concat(end(n))
      .concat('')
      .join('\n');
  },

  sing: function (first, last) {
    var i, verses = [];
    last = last || 0;

    for (i = first; i >= last; i--) {
      verses.push(this.verse(i));
    }

    return verses.join('\n');
  }

};

File: elixir/beer-song/beer_song.exs
defmodule BeerSong do
  @of_beer "of beer"
  @on_the_wall "of beer on the wall"

  @doc """
  Get a single verse of the beer song
  """
  @spec verse(integer) :: String.t
  def verse(n) do
    """
    #{n - 1 |> bottles} #{@on_the_wall}, #{n - 1 |> bottles |> String.downcase} #{@of_beer}.
    #{pass_it_or_buy_it(n - 1)}, #{n - 2 |> bottles |> String.downcase} #{@on_the_wall}.
    """
  end

  defp bottles(n) when n < 0, do: "99 bottles"
  defp bottles(0), do: "No more bottles"
  defp bottles(n), do: "#{n} bottle#{plural(n)}"

  defp plural(1), do: ""
  defp plural(_), do: "s"

  defp pass_it_or_buy_it(0), do: "Go to the store and buy some more"
  defp pass_it_or_buy_it(n), do: "Take #{take_n_down(n)} down and pass it around"

  defp take_n_down(1), do: "it"
  defp take_n_down(_), do: "one"

  @doc """
  Get the entire beer song for a given range of numbers of bottles.
  """
  @spec lyrics(Range.t) :: String.t
  def lyrics(range \\ 100..1) do
    range
    |> Enum.map(&(verse(&1)))
    |> Enum.join("\n")
  end
end
